src/python.rs

```diff
@@ -37,92 +77,95 @@ pub struct TaxonomyNode {
     parent: Option<String>,
     #[pyo3(get)]
     rank: String,
+    // Ideally this would be private
+    extra: HashMap<String, Value>,
 }
 
 #[pyproto]
 impl PyObjectProtocol for TaxonomyNode {
-    fn __repr__(&self) -> PyResult<String> {
-        Ok(format!(
-            "<TaxonomyNode (id={} rank=\"{}\" name=\"{}\"))>",
-            self.id, self.rank, self.name
-        ))
+    fn __richcmp__(&self, other: PyRef<TaxonomyNode>, op: CompareOp) -> Py<PyAny> {
+        let py = other.py();
+        match op {
+            CompareOp::Eq => (self == other.deref()).into_py(py),
+            CompareOp::Ne => (self != other.deref()).into_py(py),
+            _ => py.NotImplemented(),
+        }
+    }
+}
+
+#[pyproto]
+impl PyMappingProtocol for TaxonomyNode {
+    fn __getitem__(&self, obj: &PyAny) -> PyResult<PyObject> {
+        let key: &str = obj.extract()?;
+        let gil_guard = Python::acquire_gil();
+        let py = gil_guard.python();
+        match key {
+            "id" => Ok(self.id.to_object(py)),
+            "name" => Ok(self.name.to_object(py)),
+            "parent" => Ok(self.parent.to_object(py)),
+            "rank" => Ok(self.rank.to_object(py)),
+            _ => {
+                if self.extra.contains_key(key) {
+                    Ok(json_value_to_pyobject(self.extra.get(key).unwrap()))
+                } else {
+                    return Err(PyKeyError::new_err(format!("Key {} not found", key)));
+                }
+            }
+        }
     }
 }
 
 /// The Taxonomy object provides the primary interface for exploring a
-/// biological taxonomy. Iterating over the Taxonomy returns all the taxonomy
-/// ids in pre-order fashion and indexing into the Taxonomy object with a
-/// taxonomy id will return a tuple of the name and rank of that id.
+/// biological taxonomy.
 #[pyclass]
 #[derive(Debug)]
 pub struct Taxonomy {
```

Comment:
Not sure if this is worth it but maybe naming this `PyTaxonomy` or something could be nice for easier grepping of the codebase given that there's the `Taxonomy` trait too.

Reply:
I'm not sure, all the Python stuff is in a single file so that's probably ok to keep the "correct" name